### ISSUE TITLE
Investigate SAE-mediated world model coherence enhancement algorithm

### ISSUE THREAD

**Issue by speediedan** [speediedan/interpretune#103]
Possible SAE-mediated world model coherence engendering algorithm:
- Generate a CoT trace, then, using (perhaps first fine-tuning) an SAE 'truth' feature as a simple proxy for internal world model coherence, calculate the gradient wrt the maximum activation of the truth feature for each CoT step (ideally do so efficiently for all steps at once since it's costly).
- Next for steps with a certain threshold incoherence (gradient wrt truth as delta), generate (very modest sized internal beam search) alternate possible steps using the candidate CoT trace with the highest likelihood, proceed with backprop. For backprop, use the normal loss signal from the zero shot classification loss function, except include in that loss fn as well a term (and therefore subsequent gradient flow) that is the sum (possibly weighted by depth) of the calculated step-wise gradients (again, wrt the 'truth' feature) across the SAE layer (possibly multiple in the future) and then the remaining (or at least some portion of) the remaining parameters (possibly fts controlled).
- Other terms may need to be experimented with but keep it as simple to start for a poc.
- See if world model coherence and task performance can be improved for this simple case to start, may need to look at meta latents, a composition of more abstract coherence-related latents.
- After identifying basic latents, create an activation cache for the entire val dataset and look at max activating examples.
- Consider using the LLM to explain its entailment classification for highly confident cases versus less confident cases as a way to supplement training data and amplify coherence signal.
- Also consider using generated explanations of coherence and non-coherence (both for each example) for highly confident correct cases as a means for exposing/identifying relevant abstract meta-latents.
- Also perhaps ask the model to provide the start/stop indices in the premise and hypothesis associated with the support or contradiction to allow more granular latent probing (this is where a dataset like FEVER with similar existing data would be useful).
- Try running val dataset and capturing max activated latents, then finetuning with SAE gradients enabled (and bootstrapped initial coherence proxy loss term and prob SAE sparsity encouraging loss term), then re-running with original SAE to see which latents changed the most (diff approach to diff-SAE).
